### PR TITLE
Added CurvatureWeighting to improve Motorcycle routing.

### PR DESCRIPTION
**Pull request by boldtrn**
We created a weighting that creates fun routes for riding with a motorcycle (at least what is fun to me, when riding the motorcycle). Unfortunately we have no clue on how to create UnitTests for the routing, therefore no tests were added.

Thins that would be also nice to add are:
- Consider elevation. Roads that include elevation are usually fun to ride with the motorcycle.
- Allowing the user to create a list of nice roads that are preferred.

Also I modified the MotorcycleFlagEncoder in a way that improved the routing in my opinion. This PR belongs to #425

### CODE REVIEW

File: core/src/main/java/com/graphhopper/routing/util/CurvatureWeighting.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    {
+        super(flagEncoder, pMap);
+        this.flagEncoder = flagEncoder;
+        this.nodeAccess = ghStorage.getNodeAccess();
+    }
+
+    @Override
+    public double getMinWeight( double distance )
+    {
+        return 0.001 * distance;
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/CurvatureWeighting.java]: This is probably not correct. It needs to be a good approximation to make A* working fast and correct. If it is too small A* will be too slow, if it is bigger then a single value in real world then route results can be incorrect.

File: core/src/main/java/com/graphhopper/routing/util/CurvatureWeighting.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    /**
+     * A really small bendiness or a bendiness greater than 1 indicates an error in the calculation. Just ignore them.
+     * We use bendiness > 1.2 since the beelineDistance is only approximated,
+     * therefore it can happen on straight roads, that the beeline is longer than the road.
+     */
+    protected double correctErrors( double bendiness )
+    {
+        if (bendiness < 0.01 || bendiness > 1.2)
+        {
+            logger.info("Corrected a bendiness of: " + bendiness);
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/CurvatureWeighting.java]: no logs in Weighting please :)

Either throw an exception, add more unit tests or make it working all the time. There are really rare cases where none of them is possible.

**Comment by karussell** [comment]
Would you mind to look into the failing integration tests? If the new results are indeed an improvement?

Also I would not want to have the expensive calculations for the distance in the Weighting class, can you check if you do them before and store them somewhere how much faster this is? I would be fine with it if it is like only 10% worse

**Comment by karussell** [comment]
If you want to fine tune the routing I recommend to specify some external variable like `bendiness_booster` which you can e.g. pass through the constructor and take from the URL parameter, which influences the weighting. GraphHopper passes it from the URL in the PMap and you can easily grab like [so](https://github.com/graphhopper/graphhopper/blob/master/core/src/main/java/com/graphhopper/routing/util/FastestWeighting.java#L48)

BTW: The calcWeight method should be **very** fast due to two reason
 * faster execution speed for none-CH and CH
 * faster CH preparation

So precalculate as much as possible and e.g. leave only some external variables open which configures the final routing.

> Consider elevation. Roads that include elevation are usually fun to ride with the motorcycle.

Let us create a new issue for this as it could be a bit more effort. But you could check if elevation is available let it influence the priority. Is up and downhill different? If we can avoid this it would save us lots of resources (RAM) and make it easier as both weights would be identical. Otherwise a similar approach like the one in bike2 would be necessary: like [here](https://github.com/graphhopper/graphhopper/blob/master/core/src/main/java/com/graphhopper/routing/util/Bike2WeightFlagEncoder.java#L68) but other changes are then required like described [the docs](https://github.com/graphhopper/graphhopper/blob/master/docs/core/create-new-flagencoder.md#different-forward-and-backward-weights)

**Comment by boldtrn** [comment]
Hi peter,

I just moved the bendiness calculation to the MotorcycleFlagEncoder, probably still some work to do, but it works (at least on my machine ;)) and it should be better than the previous version. However, I am not sure if its a good idea to add it to the MotorcycleEncoder or if creating a new Encoder would be better?

I will look into the height thing in the future. In general driving upwards is preferred by most motorcyclists, but I think we should not care about that currently. Any height changes are preferred to driving horizontally. So this could be a second parameter of the motorcycle encoder. Thanks for the references to the docs.

The `booster` is a good idea. That should be considered in the future.

**Comment by karussell** [comment]
Thanks, would you also have a look into the failing tests? (use `mvn clean test verify` or run the RoutingAlgorithmIT, IT stands for integration test)

I would for now make the bendiness only available to motorcycle, maybe you/we find a way later to extract and reuse it elsewhere.

**Comment by boldtrn** [comment]
Hi Peter,

yes I know about the tests. The errors occur because I changed the speed and priority of some streets. Therefore the expected values differ from the actual values. Are you fine when I modify the  tests or should I rather reset the speed and priorities of the streets?

**Comment by karussell** [comment]
After verifying the new routes e.g. via starting the UI with the areas locally I'm fine with resetting the expected values.

> So if I understand you correctly, I shall also add a similar Test for the CurvatureWeighting, correct?

Yes, best would be an integration test (but we already have it when using the new weighting?) and one or two unit tests without import requirements just set some bendiness values in the edge and assert the weighting and test(s) for the bendiness calculation itself

**Comment by boldtrn** [comment]
Hi Peter,

I just fixed the Tests and basically copied the fastest test to use it for the curvature weighting, which produces the same results. 

> one or two unit tests without import requirements just set some bendiness values in the edge and assert the weighting and test(s) for the bendiness calculation itself

Alright, I can do something like this as well.


Best,
Robin

**Comment by boldtrn** [comment]
Hi Peter,

I just added the required test. But I had to copy the `initExampleGraph` method from the `Bike2WeightFlagEncoderTest`. The method should be probably added to some `TestHelper` but I am not sure about the appropriate position. This is probably something you should decide ;)

File: core/src/main/java/com/graphhopper/GraphHopper.java [hunk 3]
```diff
@@ -940,9 +940,16 @@ public Weighting createWeighting( WeightingMap weightingMap, FlagEncoder encoder
                 return new PriorityWeighting(encoder, weightingMap);
             else
                 return new FastestWeighting(encoder, weightingMap);
+        } else if ("curvature".equalsIgnoreCase(weighting) || weighting.isEmpty())
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/GraphHopper.java]: weighting.isEmpty should be removed here (and is already handled one if before) ... I make this in the squashed commit ...

**Comment by karussell** [comment]
Thanks again, this looks good. 

> But I had to copy the initExampleGraph method from the Bike2WeightFlagEncoderTest.

This is okay for now I think.

But I cannot create a squashed commit, I'm not that experienced with git. Can you somehow rebase the changes so that we have only one commit for all the changes? Or e.g. extract a patch and use that for one commit?

Also the germany.c_300.kml should be probably removed from the pull request as I cannot find the usage of it (?)

**Comment by boldtrn** [comment]
> Also the germany.c_300.kml should be probably removed from the pull request as I cannot find the usage of it (?)

Removed

> But I cannot create a squashed commit, I'm not that experienced with git. Can you somehow rebase the changes so that we have only one commit for all the changes? Or e.g. extract a patch and use that for one commit?

It should be possible by just click on the Auto-Merge button in github?

**Comment by karussell** [comment]
> It should be possible by just click on the Auto-Merge button in github?

No idea, I only use command line git. I could probably do it via extracting a patch out of this PR but then you won't show up as the author which is far from what I want :)

... but maybe I found a [different simpler solution](http://stackoverflow.com/questions/5308816/how-to-use-git-merge-squash)

**Comment by karussell** [comment]
Okay via the following author fake it was possible and very easy:
```bash
git merge --squash issue_534
git add .
git commit --author "your name <your@ema.il>" -m "improve ..."
```

So thanks again - I'll keep it open until successful deployment in a few days

**Comment by devemux86** [comment]
@karussell nice Git hint!

**Comment by boldtrn** [comment]
@karusell I am a terminal user as well. But all prs that are automatically mergeable, can be merged by github automatically. You just need to click on the green auto merge button in the bottom of the pr. So no terminal magic is required at this point. After merging you can also delete the branch. Next time you pull from master the branch is merged to the master.

**Comment by karussell** [comment]
The problem with this button is the clumsy history. For one PR I prefer just one commit :) ... although I have to admit that single commits from my side aren't as clean as I want them to ;)

**Comment by karussell** [comment]
This should be live now. But **fastest** weighting only, due to current small limits of multiple vehicles and CH. See #493.

As contributor I've added @boldtrn someone else?

**Comment by boldtrn** [comment]
Hi Peter,

sorry for the late reply. Sounds great :) @seeebiii and @ChristianSeitzer  were part of the team as well. If you don't mind it would be nice if you could add them as well :)

Hopefully I will find some time in several weeks to improve the curvature weighting as discussed above.

Best,
Robin

**Comment by karussell** [comment]
Thanks to them as well - added them!

**Pull request closed without merging**